### ISSUE TITLE
Support reconfigurable modules via "@forward ... with"

### ISSUE THREAD

**Issue by mirisuzanne** [sass/sass#2744]
*Edited by @nex3*

- [x] [Proposal](https://github.com/sass/sass/blob/master/accepted/forward-with.md)
- [x] [Specs](https://github.com/sass/sass-spec/pull/1499)
- [x] [Documentation](https://github.com/sass/sass-site/issues/383)
- [x] [Dart Sass](https://github.com/sass/dart-sass/issues/846)
- [ ] [LibSass](https://github.com/sass/libsass/issues/3004)

The current plan is to add a `with` clause to `@forward`. This would work like the same clause in `@use`, except that unconfigured variables or variables set with the `!default` flag would still be configurable by the modules loading the module in which the `@forward` appears. For example:

```scss
@forward "unopinionated" with (
  $hue: 0 !default, // Can be overridden by upstream users.
  $saturation: 50% // Cannot be overridden by upstream users.
);
```

---

I've been moving my Sass packages over to the module system, but I keep hitting a problem I can't figure out. I usually have a central set of configuration variables that need to:

- be configurable by users
- be used by all the internal modules

But any time I make an internal file available to other files, I lose the ability to take user-configurations:

```
Error: This module was already loaded, so it can't be configured using "with".
```

I've tried adding duplicate `!default` settings in each file that needs, but then I get:

```
Error: Module sass/_config.scss and the new module both forward a variable named $setting.
```

As far as I can tell, everything that uses a particular default variable has to live in the same file?

Is this how it's meant to work? How would you recommend handling global package config shared among partials?

**Comment by nex3** [comment]
Can you give a code example? Whatever's going on here may depend on the specifics of how you're defining and forwarding configuration.

**Comment by mirisuzanne** [comment]
I'm trying to basically set it up the way you demonstrate in the proposal document, but there's a layer missing from that sample which I'm not sure how to handle:

```scss
// colors/_config.scss
$colors: 6 !default;

// colors/_generate.scss
@use 'config';
@mixin generate-colors($colors: config.$colors) { /* mixin-logic */ }

// colors/_index.scss
@forward 'config';
@forward 'generate';

// _user.scss
@use 'colors' with ( $colors: 4 );
@include colors.generate();
```

Both the `_config.scss` variables and `_generate.scss` mixins need to be forwarded to the user – but one relies on using the other.

**Comment by nex3** [comment]
You've found a bug! It looks like it's even a bug in the spec: when you load a module with a config and that module forwards *multiple* modules, the config is passed as-is to all those modules, which will basically never do what you want. We need to be more sophisticated about how we handle that interaction.

**Comment by nex3** [comment]
Actually, it's a little bit more complicated than that. What you're seeing here specifically is actually *just* a bug in Dart Sass. According to the spec, `colors/config` is loaded twice, first with a configuration (`user` -> `colors` -> `colors/config`) and second without a configuration (`user` -> `colors` -> `colors/generate` -> `colors/config`). Even in the spec as written, this shouldn't cause an error, because only the first load has a configuration. It does in Dart Sass because it's somehow being set to an empty map instead of `null`, which is what we check for.

The spec is actually more sophisticated about this than I gave it credit for. Although it does pass the configuration to each forwarded module, it has a specific clause at the beginning of [Executing Files](https://github.com/sass/sass/blob/master/accepted/module-system.md#executing-files) to skip the check for extra configured variables when passing a configuration through `@forward`s. So even though every forwarded module receives the full configuration, they won't complain about having non-existent variables configured.

**However**, this still produces some strange edge cases:

* If you module A forwards module B, and A is loaded with config, B counts as being loaded with config too—even if none of the configured variables apply to B in practice. So if B had been loaded earlier, it will fail now.

* If module A forwards module B twice, even through different transitive paths, neither A nor B can ever be configured, because otherwise B will be loaded with configuration twice and that's forbidden.

These are fairly narrow edge cases, and I don't know whether they're worth fixing before we launch the module system, considering that it's way easier to make errors stop erroring than the reverse. But it's worth thinking about.

**Comment by mirisuzanne** [comment]
If I understand those edge-cases right, both should have workarounds – not too different from managing load-order in the current versions of Sass. I think you're right that we can fix the bug and put off the edge-cases, as long as we document the workarounds (implicitly if not explicitly).

I'm happy to help on the documentation front. I'm planning a blog post about how this all works, as I move my packages over to the new syntax.

**Comment by nex3** [comment]
The Sass team discussed this, and we've decided to move ahead with `@forward with`, [as I described here](https://github.com/sass/sass/issues/2742#issuecomment-534673043).

While talking it over, realized that it's a good practice to always write `@forward`s before the corresponding `@use`s if both exist so that the forwarded modules will be configurable by anyone loading the intermediate module. Given that, `@use with` *never* makes sense for a forwarded module, so `@forward with` is necessary (and doesn't just narrowly target the reconfigurability use-case).

**Comment by diondree** [comment]
This would also be helpful for my use case. I'm trying to create a css framework that is built ontop of bootstrap. So the idea is to extend and override the library. So I'd like to be able do the following for example:
```scss
//bootstrap.scss

//imported from node_modules
@forward 'bootstrap' with ($blue: #0366d6);
```

**Comment by nex3** [comment]
The proposal has landed! I'm going to give it a couple weeks to accumulate feedback, and then move forward on implementation.

**Comment by nex3** [comment]
Since no one has raised any issues, I'm officially declaring the proposal accepted.

**Comment by chris-dura** [comment]
I think I've encountered a similar (I hope 😬) issue, but I'm still early days in playing with the new module system, and I'm having a hard time conceptualizing if this proposal will help me to what I need to.

As a library author, I want to provide some global "feature flags" (in the form of scss `$variables`), so that my consumers can opt-out of compiling certain CSS styles, if they want. This was pretty straight forward (pun intended) using the old-school `@import`s, and I think I've successfully figured out what a _module_ consumer that uses `@use/@forward` needs to do.

However, I'm having a hard time getting it to work if my legacy consumers are still using `@import`, which I'm sure most people will be for quite awhile yet...

I threw up [an example repo](https://github.com/chris-dura/sass-use-example/blob/master/src/legacy/legacy.scss) reproducing how a legacy consumer might try to `@import` and override my library (written with `@use/@forward`).

Basically, I would expect [the output](https://github.com/chris-dura/sass-use-example/blob/master/src/legacy/legacy.css) to be:
```css
.foo {
  margin: 5px;
  color: red;
}

.baz {
  color: red;
}
```
But, it doesn't seem to be taking the variable override. ☹️

**Comment by mirisuzanne** [comment]
@chris-dura that sounds more like the bug in issue #2772 - which looks like it has a resolution implemented in dart-sass, but is not yet released.

**Comment by chris-dura** [comment]
@mirisuzanne -- thank you, I agree, that may solve it.

**Comment by roydukkey** [comment]
Does this example exhibit the same issue as described here?

```sass
// console.sass
$context: null !default

@mixin warn($message)
  @if $context != null
    $message: "'#{$context}': #{$message}"
  @warn $message

// test.sass
@use 'console'
@include console.warn('Test One') //=> "Test One"
@import 'a'

// a.sass
@use 'console' with ($context: 'a.sass')
@include console.warn('Test Two') //=> "'a.sass': Test Two"

// Terminal Error
: Test One
    test/console.sass 6:2  warn()
    test/index.sass 10:1   root stylesheet

 This module was already loaded, so it can't be configured using "with".
      ┌──> test/a.sass
    1 │ @use 'console' with ($context: 'a.sass')
      │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ new load
      ╵
      ┌──> test/index.sass
    9 │ @use 'console'
      │ ━━━━━━━━━━━━━━ original load
      ╵
      test/a.sass 1:1       @import
      test/index.sass 11:9  root stylesheet
```

I understand there have been some suggestions for workarounds and possibly staged PRs, but I not sure how it would be possible to work around this issue. Or, if this is even the issue described here.

"sass": "^1.26.1"

**Comment by mirisuzanne** [comment]
@roydukkey no, this issue is different.

The `!default` flag and `with …` clause are designed for universal configuration of a module: something you will set once for a project. Because of that, they can only be used once, the first time a module is used. The `@forward … with …` syntax is still part of that.

For a setting that you want to change multiple times in different contexts, you can override them directly:

 ```scss
// a.sass
@use 'console';

console.$context: 'a.sass'; // override the variable
@include console.warn('Test Two') //=> "'a.sass': Test Two"
```

**Comment by 0livare** [comment]
The sass `@forward` documentation currently gives the following example under [Configuring Modules](https://sass-lang.com/documentation/at-rules/forward#configuring-modules).

```scss
// _library.scss
$black: #000 !default;
$border-radius: 0.25rem !default;
$box-shadow: 0 0.5rem 1rem rgba($black, 0.15) !default;

code {
  border-radius: $border-radius;
  box-shadow: $box-shadow;
}
```

```scss
// _opinionated.scss
@forward 'library' with (
  $black: #222 !default,
  $border-radius: 0.1rem !default
);
```

```scss
// style.scss
@use 'opinionated' with ($black: #333);
```

With this setup, should it be possible to `@use 'opinionated'` more than once?

```scss
// style2.scss
@use 'opinionated' 
```

I think this would be useful for @diondree's use case in not wanting to expose bootstrap directly to the users of his library.

But my experience with the current implementation is that you would get the following:

```
  SassError: This module was already loaded, so it can't be configured using "with".
      ┌──> style2.scss
   2  │ ┌ @forward 'library' with (
   3  │ │   $black: #222 !default,
   4  │ │   $border-radius: 0.1rem !default
   5  │ │ );
      │ └─^ new load
      ╵
      ┌──> style.scss
   2  │ ┌ @forward 'library' with (
   3  │ │   $black: #222 !default,
   4  │ │   $border-radius: 0.1rem !default
   5  │ │ );
      │ └─^ original load
```

The only workaround that I can come up for this situation would be to force the users of your library to first import your configuration file, where you [manually copy](https://github.com/sass/sass/issues/2742#issuecomment-534673043) variable values over to bootstrap, before they are allowed to import anything else from your library.

**Comment by mirisuzanne** [comment]
@zposten That looks like a bug, from what you've shown, but I'm not able to reproduce it in the latest Dart Sass: `v1.26.3`

- Do you have the latest version installed?
- Your code above isn't entirely complete, so I had to infer a few things. Can you show the full setup (maybe create a gist or small repo) with this failing on the latest version consistently?

**Comment by 0livare** [comment]
@mirisuzanne, thank you for your response.  After what feels like years of investigation, I found that [my issue](https://github.com/webpack-contrib/sass-loader/issues/820) was actually caused by webpack's sass-loader, not sass itself.  So please ignore me 🙃

**Comment by cpoftea** [comment]
> 
> 
> @roydukkey no, this issue is different.
> 
> The `!default` flag and `with …` clause are designed for universal configuration of a module: something you will set once for a project. Because of that, they can only be used once, the first time a module is used. The `@forward … with …` syntax is still part of that.
> 
> For a setting that you want to change multiple times in different contexts, you can override them directly:
> 
> ```css-scss
> // a.sass
> @use 'console';
> 
> console.$context: 'a.sass'; // override the variable
> @include console.warn('Test Two') //=> "'a.sass': Test Two"
> ```

Hmm, it didn't mentioned at sass-lang docs, but it must be, I think.

**Comment by kvanbere** [comment]
Hi,
I'm trying to do something like this:
```sass
// Between 576px and 768px
@use 'vertical-rhythm' as vr-xs with (
  $base-font-size: 12px,
  $base-line-height: 16px
)

// Between 768px and 992px
@use 'vertical-rhythm' as vr-sm with (
  $base-font-size: 14px,
  $base-line-height: 20px
)

// Between 992px and 1640px
@use 'vertical-rhythm' as vr-md with (
  $base-font-size: 16px,
  $base-line-height: 24px
)

// 1640px and above
@use 'vertical-rhythm' as vr-lg with (
  $base-font-size: 20px,
  $base-line-height: 28px
)
```
I get the same error as in OP.
Would this ticket handle my use-case, or is there already a way to achieve this?

---
Not to complicate the example, but here's the rest of the support code to illustrate the utility:
```scss
// _fluid-rhythm.scss

// declarations above...

$flr-rhythm-namespaces: (
  576px: 'vr-xs',
  768px: 'vr-sm',
  992px: 'vr-md',
  1640px: 'vr-lg'
)

$flr-rythm-funs: ();
@each $sz, $ns in $flr-rhythm-namespaces {
  $flr-rythm-funs: map.merge($flr-rythm-funs, ($sz: meta.module-functions($ns)));
}

@mixin adjust-font-size-to($to-size-rat, $lines: auto) {
  $font-sizes: ();
  $line-heights: ();

  @each $dim, $ns in $flr-rythm-funs {
    $to-size: meta.call(map.get($ns, 'font-size-for-rat'), $to-size-rat);

    @if $lines == auto {
      $lines: meta.call(map.get($ns, 'lines-for-font-size'), $to-size);
    }
    $line-height: meta.call(map.get($ns, 'rhythm'), $lines, $to-size);

    $font-sizes: map.merge($font-sizes, ($dim: $to-size));
    $line-heights: map.merge($line-heights, ($dim: $line-height));
  }

  @include pfs.poly-fluid-sizing(font-size, $font-sizes);
  @include pfs.poly-fluid-sizing(line-height, $line-heights);
}



// _poly-fluid-sizing.scss
@mixin poly-fluid-sizing($property, $map) {
  // Get the number of provided breakpoints
  $length: length(map-keys($map));

  // Error if the number of breakpoints is < 2
  @if ($length < 2) {
    @error "poly-fluid-sizing() $map requires at least values"
  }

  // Sort the map by viewport width (key)
  $map: map-sort($map);
  $keys: map-keys($map);

  // Minimum size
  #{$property}: map-get($map, nth($keys,1));

  // Interpolated size through breakpoints
  @for $i from 1 through ($length - 1) {
    @media (min-width:nth($keys,$i)) {
      $value1: map-get($map, nth($keys,$i));
      $value2: map-get($map, nth($keys,($i + 1)));
      // If values are not equal, perform linear interpolation
      @if ($value1 != $value2) {
        #{$property}: linear-interpolation((nth($keys,$i): $value1, nth($keys,($i+1)): $value2));
      } @else {
        #{$property}: $value1;
      }
    }
  }

  // Maximum size
  @media (min-width:nth($keys,$length)) {
    #{$property}: map-get($map, nth($keys,$length));
  }
}



// style.scss
@use 'fluid-rhythm' as fr;

body, p {
  @include fr.adjust-font-size-to(1);
}

h1 {
  @include fr.adjust-font-size-to(1.2);
}
```

**Comment by mirisuzanne** [comment]
@donkeybonks No – same error, but a very different use-case. The `with ( ... )` syntax is meant for one-time project-wide _initial settings_. 

- Modules are designed to be configured one time only.  
- Mixins are designed to be called many times with different settings. 

If you find yourself trying to call the same module many times with different configurations, consider using mixins instead:

```scss
 @use 'vertical-rhythm' as vr;

// name it however you want...
@include vr.build($name: xs, $breakpoint: 576px, $base-font-size: 12px, $base-line-height: 16px);
@include vr.build($name: sm, $breakpoint: 768px, $base-font-size: 14px, $base-line-height: 20px);
```

I don't know your code or goals well enough to be precise, but hopefully that gives you enough to go on. You could also call that mixin repeatedly from a loop, if you want to store those values in a map…

You can also wrap all of that up in a module if you want to pass in a single nested-map configuration, and do the looping internally:

```scss
 @use 'vertical-rhythm' as vr with (
  $sizes: (
    xs: (
      base-font-size: 12px,
      base-line-height: 16px,
    ),
    sm: (
      base-font-size: 14px,
      base-line-height: 20px,
    ),
  ),
);
```

**Comment by mangelozzi** [comment]
> * Modules are designed to be configured one time only.

I think it would be great if this was added the documentation. I thought everytime one `@use` a module it was loading it in a unique way as specified with `with`.

**Comment by nex3** [comment]
> I think it would be great if this was added the documentation. I thought everytime one `@use` a module it was loading it in a unique way as specified with `with`.

That's a good point. Can you file an issue in https://github.com/sass/sass-site, or better yet a pull request to add that documentation?

**Comment by nex3** [comment]
Closing this out because [LibSass is now deprecated](https://sass-lang.com/blog/libsass-is-deprecated) and we aren't expecting to add any additional features to it.
